core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
     }
   }
 
+  /**
+   * Helper function called in ExtendedAliasExpander (used to expand identifiers in the Select list,
+   * Where, and On clauses) and ExtendedExpander (used by having/group by clauses).
+   *
+   * Finds expressions that alias to "name", in the select list.
+   *
+   * @param name The alias to search for in the select list
+   * @param nameMatcher The nameMatcher used to check if the aliases match.
+   * @param select The select to search.
+   * @param maxIdx The maximum idx into the select list to search.
+   * @return A nested pair containing the number of occurrences of the alias, one of the expressions
+   * to which the alias evaluates (if there are multiple), and the index into the select list that
+   * said expression occurs at. The information is arranged (numOccurrences, (expr, expr_idx))
+   */
+  protected static Pair<Integer, Pair<@Nullable SqlNode, Integer>> findAliasesInSelect(
+      String name, SqlNameMatcher nameMatcher, SqlSelect select, Integer maxIdx) {
+    SqlNode expr = null;
+    int numOccurrencesOfAlias = 0;
+    int idxOfOccurrence = 0;
+
+    for (int i = 0; i < maxIdx; i++) {
+      SqlNode s = select.getSelectList().get(i);
+      final String alias = SqlValidatorUtil.getAlias(s, -1);
+      if (alias != null && nameMatcher.matches(alias, name)) {
+        expr = s;
+        idxOfOccurrence = i;
```

Comment:
What happens if this matches more than once? Is it not a problem if idxOfOccurrence gets clobbered?

Reply:
An example would also be really helpful here.